### PR TITLE
[Merged by Bors] - feat(Algebra/Polynomial/Laurent): evaluate at a unit

### PR DESCRIPTION
**Pull request by ScottCarnahan**
This PR introduces a `LaurentPolynomial.smeval` function that evaluates a Laurent polynomial with coefficients in a semiring `R` at a unit in a semiring `S` that has scalar multiplication by `R`. The standard use case is when `R` is commutative and `S` is an `R`-algebra.

---

There is some conceptual overlap with #19573 which defines `lift` as a ring homomorphism to a CommRing.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Polynomial/Laurent.lean [hunk 1]
```diff
@@ -544,4 +551,78 @@ lemma toLaurent_reverse (p : R[X]) :
...
+    rw [mul_add, eval_add, eval_add, smul_add, hp, hq]
+  | h_C_mul_T n s =>
+    rw [← mul_assoc, ← map_mul, eval_C_mul_T_n, eval_C_mul_T_n, mul_smul]
+
+/-- Evaluation as an `R`-linear map. -/
+@[simps]
+def leval (R) [Semiring R] [Module R S] (x : Sˣ) : R[T;T⁻¹] →ₗ[R] S where -- make R explicit?
+    toFun f := f.eval x
+    map_add' f g := eval_add f g x
+    map_smul' r f := by simp [smul_eq_C_mul]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Polynomial/Laurent.lean:622]:
```suggestion
  toFun f := f.eval x
  map_add' f g := eval_add f g x
  map_smul' r f := by simp [smul_eq_C_mul]
```

File: Mathlib/Algebra/Polynomial/Laurent.lean [hunk 2]
```diff
@@ -544,4 +551,78 @@ lemma toLaurent_reverse (p : R[X]) :
...
+@[simp]
+theorem eval_C (r : R) : (C r).eval x = r • 1 := by
+  rw [← single_eq_C, eval_single x (0 : ℤ) r, zpow_zero, Units.val_one]
+
+end SMulWithZero
+
+section MulActionWithZero
+
+variable [Semiring R] [AddCommMonoid S] [MulActionWithZero R S] [Monoid S] (f g : R[T;T⁻¹])
+(x y : Sˣ)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Polynomial/Laurent.lean:588]:
```suggestion
  (x y : Sˣ)
```

File: Mathlib/Algebra/Polynomial/Laurent.lean [hunk 3]
```diff
@@ -544,4 +551,78 @@ lemma toLaurent_reverse (p : R[X]) :
...
+theorem eval_C_mul (r : R) : (C r * f).eval x = r • (f.eval x) := by
+  induction f using LaurentPolynomial.induction_on' with
+  | h_add p q hp hq=>
+    rw [mul_add, eval_add, eval_add, smul_add, hp, hq]
+  | h_C_mul_T n s =>
+    rw [← mul_assoc, ← map_mul, eval_C_mul_T_n, eval_C_mul_T_n, mul_smul]
+
+/-- Evaluation as an `R`-linear map. -/
+@[simps]
+def leval (R) [Semiring R] [Module R S] (x : Sˣ) : R[T;T⁻¹] →ₗ[R] S where -- make R explicit?
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Polynomial/Laurent.lean:619]:
```suggestion
variable (R) in
/-- Evaluation as an `R`-linear map. -/
@[simps]
def leval : R[T;T⁻¹] →ₗ[R] S where
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
Stylistic comments, no opinion either way on the contents

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**